### PR TITLE
Refactor Python and Cython internals for groupby aggregation

### PR DESCRIPTION
**Pull request by vyasr**
This PR makes some improvements to the groupby/aggregation code that I identified while working on #7731. The main purpose is to make the code logic easier to follow and reduce some unnecessary complexity; I see minor but measurable performance improvements (2-5% for small datasets) as well, but those are mostly just side effects here. Specifically, it makes the following changes:

1. Inlines the logic for dropping unsupported aggregations. The old function was repetitive and necessitated looping over the aggregations twice, whereas the new approach drops unwanted aggregations on the fly so it only loops once. The new code also makes it so that you only construct a C aggregation object once.
2. Merges the logic from `_AggregationFactory` into `Aggregation`, and removes the constructor for `Aggregation`. The one downside here is that the Cython `Aggregation` object's constructor no longer places it in a valid state; however, in practice the object is always constructed via either the `make_aggregation` function or its various factories, and the object's constructor was only every used in `_drop_unsupported_aggs` anyway. The benefit is we remove the fragmentation between these two classes, making the code much more readable, and the `Aggregation` class actually serves a purpose now beyond just providing a single property `kind` that is only used once: it is now the primary way that other Cython files interact with aggregations. This also means that in most places other Cython modules don't need to work with `unique_ptr[aggregation]` as much anymore (although they do still have to move `Aggregation.c_obj` for performance reasons). `make_aggregation` now returns the Cython class instead of the underlying C++ one.
3. Modified all the "allowed aggregations" sets to use the uppercase names of the aggregations. In addition to simplifying the code a tiny bit, this helps reduce confusion between the aggregation names used in Python for pandas compatibility and the libcudf names (for instance, `idxmin` vs `argmin`, now `ARGMIN`).
4. Explicitly defines all the aggregations on a groupby. I discussed this briefly with @shwina, the change has pros and cons. The benefit is that all of these methods are properly documented now, there's less magic (the binding of methods to a class after its definition can be confusing for less experienced Python developers and has a lot of potential gotchas), and we can use the simpler string-based agg definition wherever possible. The downside is that we now have to define all of these methods. I think the change is definitely an improvement, but I'm happy to change it back if anyone can suggest a better alternative. In the long run we probably need to find a way to share both code and docstrings more effectively between all aggregations (DataFrame, Series, and GroupBy).

### CODE REVIEW

**Comment by vyasr** [comment]
rerun tests

File: python/cudf/cudf/_lib/aggregation.pxd
Comment (shwina) [inline python/cudf/cudf/_lib/aggregation.pxd:11]: Any `cdef` functions that return a non-Python object have no way of communicating exceptions to their callers unless you mark them `except *` or `except ?` (the latter when you can specify a specific return value).

It's a good idea to do this even if the underlying C++ code is "guaranteed" not to raise any exceptions because Python can very easily sneak into the implementation, or there may be other errors like `RecursionError` that can originate from inside the function.

Comment (vyasr) [reply python/cudf/cudf/_lib/aggregation.pxd:11]: Thanks for finding, I looked into the syntax but forgot to revisit. I was running into some issues with this syntax because of the value of the return type, but perhaps at this stage I've changed enough that it will just work.

Comment (shwina) [reply python/cudf/cudf/_lib/aggregation.pxd:11]: I don't think we need to do `except *` anymore, now that we're returning a Python object. Just commented as an FYI :)

Comment (vyasr) [reply python/cudf/cudf/_lib/aggregation.pxd:11]: Ah. The reason this gave errors is that Cython classes are Python types, in the sense that they are still represented by `PyObject *` pointers, which means that error propagation can proceed via the normal route in CPython: if a function returns NULL, the calling code needs to check for exceptions. Before, we were actually returning a true C(++) object, so this wasn't the case. We don't need this anymore.

Comment (vyasr) [reply python/cudf/cudf/_lib/aggregation.pxd]: I wish threads would update without me refreshing the page :) thanks for confirming

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

File: python/cudf/cudf/_lib/groupby.pyx
Comment (shwina) [inline python/cudf/cudf/_lib/groupby.pyx:130]: Given that `dtype` is guaranteed to be an instance of a known type, do you think we can replace this with a lookup from a dict with types (such as `ListDtype`) as keys?

File: python/cudf/cudf/core/groupby/groupby.py
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:572]: I'm +-1 on these changes where I don't think the previous method of dynamically adding attributes to reduce repetition was too bad. At the same time I agree this is easier to understand and lets us document aggregations better.

Another option is to use descriptors although I'd consider that even more magical:

```python
class GroupBy:
    sum = AggregationMethod("sum")  # adds a `sum` method and grabs docstring from `DataFrame.sum`?
    idxmin = AggregationMethod("idxmin")
    ...
```

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:130]: How would you map the boolean outputs of the various `is_*` functions to the keys?

File: python/cudf/cudf/core/groupby/groupby.py
Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:572]: I think this is probably something to just get some opinions on. I try to avoid magic when it's not needed to make it easier for more devs, and here we also gain docs, but I'm fine with whatever solution we choose. Maybe I'll just post in Slack later and get some opinions.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (shwina) [reply python/cudf/cudf/_lib/groupby.pyx:130]: I guess I was picturing something like:

```python
dtype_to_aggs = {ListDtype: _LIST_AGGS, np.dtype("object"): _STRING_AGGS, ...}
valid_aggregations = _dtype_to_aggs[dtype]
```

Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:130]: Ah I see what you mean by a known type, you can guarantee the input will be such that you won't need to worry about some of the more esoteric cases handled by these different functions. In any case, let's hold off on this and see if it can be subsumed by some of the broader dtype changes I've proposed.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:153]:
```suggestion
                    c_agg_requests.back().aggregations.push_back(
                        move(agg_obj.c_obj)
                    )
            if c_agg_requests.back().aggregations.empty():
                c_agg_requests.pop_back()
            else:
                c_agg_requests.back().values = col.view()
```

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:117]:
```suggestion
```

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:155]: is this the right error message? 
(non numeric types are also supported now and this error message is outdated)

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:150]: Instead of push_back and pop_back, create a temp and push to vector only if temp is non-empty.

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:155]: Yes, this is outdated. Perhaps something like this would be better:
> All requested aggregations are unsupported for column of the given dtypes.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:150]: I ran into issues where Cython would try and create a temporary pointer somewhere in the generated code, and since the agg is of type `unique_ptr[aggregation]` that is an invalid copy construction and fails to compile. I'll see if I can find a way around that.

Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:150]: Found the problem. In order to handle exceptions in the C++ function (requested by the `except +` in the pxd file) the Cython-generated C++ creates a temporary variable to hold the `unique_ptr[aggregation_request]` on construction, checks for an error, then assigns it to the correct variable if no error was detected. This necessitates and explicit move on construction of the agg request that wouldn't normally be necessary in C++.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:155]: It looks like this error exists to match the pandas error verbatim. We test this in `test_groupby.py::test_raise_data_error`. While this error is underspecified based on `cudf` features, I'm not sure if we should change it or if we're OK with diverging from pandas at the level of the text of the exception.

File: python/cudf/cudf/core/groupby/groupby.py
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:572]: Let's go with this for now; it will be easier to go back to monkey-patching attributes if at some point we decide that's. what we want.

File: python/cudf/cudf/core/groupby/groupby.py
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:587]: Given that we are now writing docstrings-per-aggregation, can we be more precise?

```suggestion
        """Compute the sum of the values in each group."""
```

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:587]: ditto for other aggs

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:587]: Sounds good, I'll update these and add return values. I also went ahead and posted in Slack to see if anyone has opinions before we finalize.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:154]: @kkraus14 @shwina do we need to match pandas verbatim here, or is it sufficient to throw the same exception type but have a different message? We support meaningful aggregations on non-numeric types, but this error alone could be misleading and suggest otherwise.

Comment (shwina) [reply python/cudf/cudf/_lib/groupby.pyx:154]: I think it's sufficient to match the error type

File: python/cudf/cudf/core/groupby/groupby.py
Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:572]: Fine with me. Apologies for not posting on this earlier, last week seemed like a pretty busy time for everyone and there didn't seem to be much urgency to get this PR reviewed/merged, but now #7759 is blocked by this. I'll follow up with making this change if we decide to switch back.

File: python/cudf/cudf/core/groupby/groupby.py
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:587]: Sorry for the runaround, but shouldn't we be explicit that this is a group-wise sum? 

```suggestion
        """Compute the sum of the values in each group."""
```

Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:587]: Yes that's a good point, but perhaps we also need to mention that it's a sum (or whatever operation) of each column per group. How about:
> Compute the column-wise sum of the values in each group.

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:587]: Sounds good to me

**Review by shwina — APPROVED** [review]
LGTM! Very nice cleanup, @vyasr!

File: python/cudf/cudf/_lib/aggregation.pyx
Comment (galipremsagar) [inline python/cudf/cudf/_lib/aggregation.pyx:292]:
```suggestion
        raise TypeError(f"Unknown aggregation {format}")
```

We can use f-strings here.

**Review by galipremsagar — APPROVED** [review]

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
Why are there only 4 checks showing in this PR?

**Comment by cwharris** [comment]
rerun tests

**Comment by vyasr** [comment]
> Why are there only 4 checks showing in this PR?

@karthikeyann all CI was blocked yesterday due to connectivity issues. The queue is slowly being emptied, I'll merge in #7968 once that's finalized then trigger tests again.

**Comment by vyasr** [comment]
I can't merge the changes from #7968 without triggering C++ reviews, so I'll just wait for that to get merged into `branch-0.20` and merge that down after.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
